src/EigenAirdrop.sol

```diff
@@ -0,0 +1,266 @@
+        }
+        if (_amountToClaim == 0 || _amountToClaim > amounts[msg.sender]) {
+            revert NoAirdrop();
+        }
+        _;
+        amounts[msg.sender] -= _amountToClaim;
+        totalAmount -= _amountToClaim;
+    }
+
+    /**
+     * @dev Disables initializers to prevent contract from being reinitialized.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the airdrop contract with the provided parameters.
+     * @param _owner The address of the owner.
+     * @param _safe The address of the safe holding the tokens.
+     * @param _token The address of the token being airdropped.
+     * @param _strategy The address of the strategy for restaking.
+     * @param _strategyManager The address of the strategy manager.
+     * @param _deadline The timestamp on which the claims are no longer valid.
+     * @param _userAmounts An array of user amounts for the airdrop.
+     */
+    function initialize(
+        address _owner,
+        address _safe,
+        address _token,
+        address _strategy,
+        address _strategyManager,
+        uint256 _deadline,
+        UserAmount[] memory _userAmounts
+    )
+        public
+        initializer
+    {
+        __Pausable_init();
+        __Ownable_init(_owner);
+        __ReentrancyGuard_init();
+
+        if (
+            _safe == address(0) || _token == address(0) || _strategy == address(0)
+                || _strategyManager == address(0) || _deadline == 0
+        ) {
+            revert InvalidInit();
+        }
+
+        safe = _safe;
+        deadline = _deadline;
+        token = IERC20(_token);
+        strategy = IStrategy(_strategy);
+        strategyManager = IStrategyManager(_strategyManager);
+
+        _updateUserAmounts(_userAmounts);
+    }
+
+    /**
+     * @notice Pauses the contract, preventing claims.
+     * Can only be called by the owner.
+     */
+    function pause() external onlyOwner whenNotPaused {
+        _pause();
+    }
+
+    /**
+     * @notice Unpauses the contract, allowing claims.
+     * Can only be called by the owner.
+     */
+    function unpause() external onlyOwner whenPaused {
+        _unpause();
+    }
+
+    /**
+     * @notice Updates user amount for the airdrop. Only callable by the owner when the contract is paused.
+     * @param _userAmounts An array of updated user amount.
+     */
+    function updateUserAmounts(UserAmount[] memory _userAmounts) external onlyOwner whenPaused {
+        _updateUserAmounts(_userAmounts);
+    }
+
+    /**
+     * @dev Internal function to update user amount and recalculate claimable amounts.
+     * @param _userAmounts An array of updated user amount.
+     */
+    function _updateUserAmounts(UserAmount[] memory _userAmounts) internal {
+        for (uint256 i; i < _userAmounts.length; i++) {
+            if (amounts[_userAmounts[i].user] > 0) {
+                totalAmount -= amounts[_userAmounts[i].user];
```

Comment:
can we gas optimize this so it's not written 2 times to storage on every loop?

Let's use one memory variable  for totalAmount

In this case this loop is very large and gas will add up

Reply:
Also let's simulate the number of entries in the sheet in the test to get a clear number of total gas

Reply:
Sure. I'll do this

Reply:
Added a gas snapshot in [a8abf92](https://github.com/yieldnest/yieldnest-eigenlayer-airdrop/pull/2/commits/a8abf926fc4589a94f92e992dee7881bed2f43fa)

> EigenAirdropTest:testUpdateUserAmountsWithSampleData() (gas: 54607069)